supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -41,6 +42,9 @@ class Sample:
     """The corresponding data about the circuit that is needed when analyzing results
     (e.g. cycle depth)."""
 
+    uuid: uuid.UUID = field(default_factory=uuid.uuid4)
```

Comment:
The `sample.circuit_index` attr is slightly different - it indicates the index of the random circuit within the given "batch" (i.e. among all circuits that share the same cycle depth). So if you have 10 different cycle depths and have chosen to `num_circuits=20` there will be 10 different circuits each with the same `sample.circuit_index` value. Perhaps we should have a better name than `circuit_index`

Reply:
ah right makes sense. One other thought: if we add something like this to `Sample`:
```python
def __hash__(self): -> int:
    return id(self)
``` 
could we just use the samples themselves as dictionary keys?

Reply:
How would this work if the user wanted to save the samples (e.g. by pickling or writing to a JSON) so that they could then re-use them later after they had run the experiments on a device? I don't think there would be a way to guarantee the `id` is the same?

This adds the extra question of how would a user save an experiment so that they can re-use later. I'm going to add some basic functionality to sort this!

Reply:
yeah good call. tbh my concern with using randomly-generated uuids is similar (and also that it creates one more variable to have to keep track of/risk messing up)

i guess one final thought: i'd think as a user the most natural way to refer to a particular sample would be its position in `experiment.samples` (and we assume this order to be fixed in run_on_device, so it seems reasonable to do the same here?). Would it make sense to use something like `sample.index` which we'd just assign consecutively?

Reply:
> This adds the extra question of how would a user save an experiment so that they can re-use later. I'm going to add some basic functionality to sort this!

one option for this is to add `_json_dict_` and `_from_json_dict_` magic methods to make these objects work with cirq's json tooling

Reply:
Ah yeah that would work! I've just pushed a version doing this semi-manually but I can switch it over tomorrow to make it use the magic methods

Reply:
> yeah good call. tbh my concern with using randomly-generated uuids is similar (and also that it creates one more variable to have to keep track of/risk messing up)
> 
> i guess one final thought: i'd think as a user the most natural way to refer to a particular sample would be its position in `experiment.samples` (and we assume this order to be fixed in run_on_device, so it seems reasonable to do the same here?). Would it make sense to use something like `sample.index` which we'd just assign consecutively?

I guess reasoning to use some form of uuid was in case there is no guarantee that the user runs the circuits in order. I suppose either way it doesn't really matter, at the end of the day we are just giving the samples some ID so we can arrange them all again later - all approaches are equivalent up to a hash-map 😂 
In the interest of reducing dependencies/imports I'll switch it over to an integer indexing tomorrow

Reply:
I've added the functionality for using cirq's json magic methods.

One thought I had overnight on the uuid vs index question is that there could be a situation where the user constructs multiple experiments (say testing different gates) and then goes away and runs them all at the same time. Using a per-experiment index would risk these indexes getting muddled whereas using a uuid means that this can't happen (I've also switched to using uuid1 by default to ensure there cannot be any collisions).

Reply:
ah that's an interesting example!

i guess i still think i'd personally prefer to use indices - it doesn't seem too hard to track which experiment my circuit belongs to, and i feel like chances are my downstream logic won't be sorted by uuids in the same way anyway - most likely i'll be taking all the circuits of a given experiment, possibly batching them, getting results back as a list/lists, and now have to use their order/indices to associate them with uuids again (which would kinda defeats their original purpose). i'd also probably want to be inspecting my results manually, which is a bit more cumbersome if they're organized by uuid

i can definitely imagine workflows for which it'd be really helpful now though. i just worry that there are possibly more cases for which it would be adding unnecessary complexity

maybe we could accept both? e.g. if the key is a string, assume it's a uuid, otherwise assume it's an index

one other thing that maybe would be helpful (possibly as a follow up): wdyt about adding something like `Experiment.__getitem__` that accepts both uuids and indices and returns the corresponding sample? if nothing else this would simplify inspection for results organized by uuid

Reply:
> I've added the functionality for using cirq's json magic methods.

nice! how hard would it be to split that change into its own pr? no worries at all if it's a lot, but if we can it'd make it a bit easier to review/look back on/etc